src/common/fail_loc.c

```diff
@@ -179,12 +179,19 @@ daos_fail_init(void)
 
 	rc = d_fault_inject_init();
 	if (rc != 0 && rc != -DER_NOSYS)
-		return rc;
+		D_GOTO(out, rc);
 
+	/* Log, but no not propagate error on registering the fault as this
+	 * leads to incorrect reference counts in daos_debug_init() and
+	 * resulting resource leaks.
+	 */
 	rc = d_fault_attr_set(DAOS_FAIL_UNIT_TEST_GROUP, attr);
-	if (rc)
-		d_fault_inject_fini();
+	if (rc != 0)
```

Comment:
we are not supposed to call d_fault_inject_fini()??
it was init earlier in the function.

Reply:
No.  This code is delicate.  From memory daos_fail_init() is ref counted and called from daos_debug_init() and daos_init(), if you call d_fault_inject_fini() here then it's properly closed but daos_debug_init() ignores the error, then daos_init() is called and re-initiaises it, at which point it works, so the first iteration of the fault injection test injects one fault (in fault_attr_set()), but then forgets it's supposed to forget one fault and carries on as normal.

I can do further testing to confirm but getting this startup code to work took a few iterations and this change was required.

Reply:
If you call d_fault_inject_fini() here then it's gets stuck in the following stack trace:
`#0  0x00007f14f1975585 in pthread_spin_lock () from /lib64/libpthread.so.0
#1  0x00007f14f043b578 in d_should_fail (fault_attr=0x24f7ee0) at src/gurt/fault_inject.c:659
#2  0x00007f14f042be25 in d_hash_table_create_inplace (feats=feats@entry=1, bits=bits@entry=13, priv=priv@entry=0x0, hops=hops@entry=0x7f14f06567e0 <fa_table_ops>, htable=htable@entry=0x7f14f0656ca0 <d_fi_gdata+64>) at src/gurt/hash.c:783
#3  0x00007f14f0439eec in d_fi_gdata_init () at src/gurt/fault_inject.c:438
#4  0x00007f14f196f20b in __pthread_once_slow () from /lib64/libpthread.so.0
#5  0x00007f14f043db5c in d_fault_inject_init () at src/gurt/fault_inject.c:477
#6  0x00007f14f06be56d in crt_init_opt (grpid=grpid@entry=0x0, flags=flags@entry=0, opt=0x0) at src/cart/crt_init.c:314
#7  0x00007f14f16509bb in daos_eq_lib_init () at src/client/api/event.c:90
#8  0x00007f14f16557c8 in daos_init () at src/client/api/init.c:184
#9  0x00007f14f27d703a in check_ioctl_on_open (fd=fd@entry=5, entry=entry@entry=0x7ffce0493e40, flags=flags@entry=0, status=<optimized out>) at src/client/dfuse/il/int_posix.c:601
#10 0x00007f14f27d81f8 in dfuse_open (pathname=0x7ffce049525a "/tmp/dfs_test/3dcb9152-1653-4111-8a07-261345887331/test_file", flags=0) at src/client/dfuse/il/int_posix.c:749
#11 0x0000000000401d55 in main ()`

Reply:
this is probably a multi-thread bug like forgetting to release a lock?
I can see in daos_debug_init and daos_init that the error is checked from daos_fail_init() so not sure what you are referring to.
from what i can tell your change introduces a leak. maybe someone else more familiar with this code path needs to review this instead of me because i don't understand how it doesn't.

Reply:
Should be fixed now.  It was a combination of d_fault_inject_fini() having a bug and the IL not checking the failure.